### PR TITLE
Add Mux and Demux to hydro

### PR DESCRIPTION
**Pull request by Squelsh**
I was missing a Demux tool. So I brought the old code back to life.
While working on that, I also fixed the service-servers of the Mux tool, so more than one Mux can be used.

Code is tested for hydro. But I think it can also be merged into newer versions. As I have no newer version running, I could not test. ==> Would be nice, if someone could do that.

If my request gets merged, I am happy to update the ROS WIKI page.

### CODE REVIEW

File: tools/topic_tools/src/demux.cpp [hunk 1]
```diff
@@ -1,9 +1,9 @@
 ///////////////////////////////////////////////////////////////////////////////
 // demux is a generic ROS topic demultiplexer: one input topic is fanned out
-// to 1 of N output topics. A service and topic subscription are provided
-// to select between the outputs.
+// to 1 of N output topics. A service is provided to select between the outputs
 //
-// Copyright (C) 2009, Morgan Quigley
+// Copyright (C) 2014, Andreas Hermann
```
Comment (tfoote) [inline tools/topic_tools/src/demux.cpp:6]: The copyright should be side by side not replace, even though it's mostly new. 

```
Copyright (C) 2009, Morgan Quigley
Copyright (C) 2014, Andreas Hermann
```

**Comment by dirk-thomas** [comment]
This comment (https://github.com/Squelsh/ros_comm/blob/hydro-devel/tools/topic_tools/src/mux.cpp#L301) states explicitly that the services are provided on the node "mux" and suggests using remapping in order to use multiple mux instances.

This patch would effectively change the public API of the tool which is not a feasible change for Hydro. Even for Indigo it is questionable if that should be changed.

The demux part is definitely a good additional. Thank you for porting this functionality.

**Comment by Squelsh** [comment]
Hi Tully, hi Dirk,
sorry for the quirks. I totally agree with your comments.
@tfoote : I corrected the Copyright as suggested.
@dirk-thomas : I reverted my changes and adapted the demux to behave the same way as the Mux.

**Comment by dirk-thomas** [comment]
I will add the missing install rule (https://github.com/ros/ros_comm/pull/407/files#diff-ae340f410bae427e1853101dfdbade7eR55) when merging it into _indigo-devel_.

**Comment by dirk-thomas** [comment]
For `mux` there are some Python scripts to ease the operations (add, delete, list, select). Wouldn't it make sense to have similar scripts for `demux`?

**Comment by Squelsh** [comment]
I did not notice these scripts till now. Will have a look on monday.

**Comment by Squelsh** [comment]
Wrong button )):

However: I don't see the point in a script, where the copyright is three times the lenghts of the script itself, but I committed the according scripts for the demux.

**Comment by dirk-thomas** [comment]
Merged to indigo-devel in 4edc26b3ab23f39073d901bf1c69066c98277e81. It will be backported to hydro-devel before the next patch release. Thank you for the PR.

**Pull request closed without merging**